src/core/annotations/qgsannotationlayer.h

```diff
@@ -0,0 +1,126 @@
+ * z-order.
+ *
+ * \since QGIS 3.16
+ */
+class CORE_EXPORT QgsAnnotationLayer : public QgsMapLayer
+{
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Setting options for loading annotation layers.
+     * \since QGIS 3.16
+     */
+    struct LayerOptions
+    {
+
+      /**
+       * Constructor for LayerOptions.
+       */
+      explicit LayerOptions( const QgsCoordinateTransformContext &transformContext )
+        : transformContext( transformContext )
+      {}
+
+      /**
+       * Coordinate transform context
+       */
+      QgsCoordinateTransformContext transformContext;
+
+    };
+
+
+    /**
+     * Constructor for a new QgsAnnotationLayer with the specified layer \a name.
+     *
+     * The \a options argument specifies load-time layer options.
+     */
+    QgsAnnotationLayer( const QString &name, const QgsAnnotationLayer::LayerOptions &options );
+    ~QgsAnnotationLayer() override;
+
+    /**
+     * Adds an \a item to the layer.
+     *
+     * Ownership of \a item is transferred to the layer.
+     *
+     * Returns the unique ID assigned to the item.
+     */
+    QString addItem( QgsAnnotationItem *item SIP_TRANSFER );
+
+    /**
+     * Removes (and deletes) the item with matching \a id.
+     */
+    bool removeItem( const QString &id );
+
+    /**
+     * Returns a map of items contained in the layer, by unique item ID.
+     *
+     * This map contains references to items owned by the layer, and ownership of these remains
+     * with the layer.
+     */
+    QMap<QString, QgsAnnotationItem *> items() const { return mItems; }
+
+    /**
+     * Sets the \a opacity for the annotation layer, where \a opacity is a value between 0 (totally transparent)
+     * and 1.0 (fully opaque).
+     * \see opacity()
+     */
+    void setOpacity( double opacity ) { mOpacity = opacity; }
+
+    /**
+     * Returns the opacity for the annotation layer, where opacity is a value between 0 (totally transparent)
+     * and 1.0 (fully opaque).
+     * \see setOpacity()
+     */
+    double opacity() const { return mOpacity; }
+
+    QgsAnnotationLayer *clone() const override SIP_FACTORY;
+    QgsMapLayerRenderer *createMapRenderer( QgsRenderContext &rendererContext ) override SIP_FACTORY;
+    QgsRectangle extent() const override;
+    void setTransformContext( const QgsCoordinateTransformContext &context ) override;
+    bool readXml( const QDomNode &layerNode, QgsReadWriteContext &context ) override;
+    bool writeXml( QDomNode &layer_node, QDomDocument &doc, const QgsReadWriteContext &context ) const override;
+    bool writeSymbology( QDomNode &node, QDomDocument &doc, QString &errorMessage, const QgsReadWriteContext &, StyleCategories categories = AllStyleCategories ) const override;
+    bool readSymbology( const QDomNode &node, QString &errorMessage, QgsReadWriteContext &context, StyleCategories categories = AllStyleCategories ) override;
+
+  private:
+    QMap<QString, QgsAnnotationItem *> mItems;
```

Comment:
What do we gain by giving the item ids such an importance (lookup table here, remove by id)... compared to using pointers as main reference?

Reply:
I was aiming for a persistent ID which could be used between project reloads/layer clones/etc, and for the extra safety that storing a string uuid gives vs storing a raw pointer to an annotation (e.g. in the drawing UI classes)

Reply:
I think I remember you saying some years ago that you wish we used more layer pointers rather than layer ids. Did that change, is the situation different here or is my memory distorted?

Reply:
I can't recall exactly, but in any case now I'd prefer pointers with the layer stored in a QPointer :wink: But that approach only works for QObjects, and I don't really want to make the annotation items qobjects themselves.

Reply:
Thanks for the clarification. I think shared_ptr don't play that well with sip ?! No more objections :+1:

Reply:
I tried a while ago to make sip handle unique ptrs with absolutely no luck.. unfortunately shared ptrs would be even harder. Maybe qt for python will help fix this in the long run...